test/bats/provider.bats

```diff
@@ -63,92 +78,141 @@ teardown(){
+    [[ "$result" == "hello-sync1" ]]
+    result=$(kubectl --namespace=test exec $POD -- printenv | grep SECRET_2_PWD | awk -F"=" '{ print $2 }' | tr -d '\r\n')
+    [[ "$result" == "hello-sync2" ]]
+}
+
+@test "5 SecretProviderClass with query parameters and PUT method" {
+    kubectl --namespace=test apply -f $CONFIGS/nginx-pki.yaml
+    kubectl --namespace=test wait --for=condition=Ready --timeout=60s pod nginx-pki
+
+    result=$(kubectl --namespace=test exec nginx-pki -- cat /mnt/secrets-store/certs)
+    [[ "$result" != "" ]]
+
+    # Ensure we have some valid x509 certificates.
+    echo "$result" | jq -r '.data.certificate' | openssl x509 -noout
+    echo "$result" | jq -r '.data.issuing_ca' | openssl x509 -noout
+    echo "$result" | jq -r '.data.certificate' | openssl x509 -noout -text | grep "test.example.com"
+}
+
+@test "6 Dynamic secrets engine, endpoint is called only once per SecretProviderClass" {
+    # Setup postgres
+    POSTGRES_PASSWORD=$(openssl rand -base64 30)
+    kubectl --namespace=test create secret generic postgres-root \
+        --from-literal=POSTGRES_USER="root" \
+        --from-literal=POSTGRES_PASSWORD="${POSTGRES_PASSWORD}"
+    kubectl --namespace=test apply -f $CONFIGS/postgres.yaml
+    kubectl wait --namespace=test --for=condition=Ready --timeout=5m pod postgres
+    POSTGRES_POD_IP=$(kubectl --namespace=test get pod postgres -o json | jq -r '.status.podIP')
+
+    # Configure vault to manage postgres
+    kubectl --namespace=csi exec vault-0 -- vault secrets enable database
+    kubectl --namespace=csi exec vault-0 -- vault write database/config/postgres \
+        plugin_name="postgresql-database-plugin" \
+        allowed_roles="*" \
+        connection_url="postgres://{{username}}:{{password}}@${POSTGRES_POD_IP}:5432/db?sslmode=disable" \
+        username="root" \
+        password="${POSTGRES_PASSWORD}"
+    cat $CONFIGS/postgres-creation-statements.sql | kubectl --namespace=csi exec -i vault-0 -- vault write database/roles/test-role \
+        db_name="postgres" \
+        default_ttl="1h" max_ttl="24h" \
+        creation_statements=-
+
+    # Now deploy a pod that will generate some dynamic credentials.
+    kubectl --namespace=test apply -f $CONFIGS/nginx-dynamic-creds.yaml
+    kubectl --namespace=test wait --for=condition=Ready --timeout=60s pod nginx-dynamic-creds
+
+    # Read the creds out of the pod and verify they work for a query.
+    DYNAMIC_USERNAME=$(kubectl --namespace=test exec nginx-dynamic-creds -- cat /mnt/secrets-store/dbUsername)
+    DYNAMIC_PASSWORD=$(kubectl --namespace=test exec nginx-dynamic-creds -- cat /mnt/secrets-store/dbPassword)
+    result=$(kubectl --namespace=test exec postgres -- psql postgres://${DYNAMIC_USERNAME}:${DYNAMIC_PASSWORD}@127.0.0.1:5432/db --command="SELECT usename FROM pg_catalog.pg_user" --csv | sed -n '3 p')
+
+    [[ "$result" != "" ]]
+    [[ "$result" == "${DYNAMIC_USERNAME}" ]]
 }
```

Comment:
nit: can you add a test for a SecretProviderClass that mixes different secret engines? For example let's say I have an app that needs database access, certs and KV. A single secret provider class should have all the secrets my app needs in a single object. Just want to make sure we have this case covered in the tests.

Reply:
Good idea, done in the next PR.